### ISSUE TITLE
Consider inferring types for un-annotated function parameters with defaults

### ISSUE THREAD

**Issue by carljm** [astral-sh/ty#3251]
For this function:

```py
def f(x="foo"): ...
```

We currently allow any actual argument to be passed to the `x` parameter, since it has no annotated type. And internally to the function, we consider the type of `x` to be `Unknown | Literal["foo"]`. This is a consistent and reasonable approach that follows the gradual guarantee.

Pyright takes a different approach; it infers a "declared" type of `str` for the `x` parameter, based on the signal from the default value. So it will error on the call `f(1)`, and it treats `x` internally as of type `str`.

Pyright's approach is parallel to how we handle attributes, after https://github.com/astral-sh/ruff/pull/24531. It risks some false positives on untyped code (by not following the gradual guarantee), but probably in most cases gives better/more type information on untyped code.

I think we could consider using Pyright's approach here. It would be interesting to see what the ecosystem fallout looks like.

**Comment by sharkdp** [comment]
This was discussed internally. We plan to do this, unless the ecosystem report tells us something that we have not anticipated.

**Comment by carljm** [comment]
I think we would need to "promote" `None` to `Unknown | None` (as we do for attributes) for this to be workable.

Not sure this is a stable blocker, though it's a significant change to semantics, so would be nice to get done before stable.

I feel like I have a vague memory of someone trying this out and reporting problematic cases from the ecosystem, but I don't see any evidence of that attached to this PR...

**Comment by AlexWaygood** [comment]
> I feel like I have a vague memory of someone trying this out and reporting problematic cases from the ecosystem, but I don't see any evidence of that attached to this PR...

I think it was https://github.com/astral-sh/ruff/pull/24943. And IIRC, as well as the `None` issue there were also _lots_ of parameters that were plainly expected to take `int`s or `float`s but had `int` defaults, so started emitting false positives in the ecosystem report.

**Comment by sharkdp** [comment]
I did some more experiments here and I am now suggesting that we *do not* implement this proposal. See PR description in https://github.com/astral-sh/ruff/pull/24943 for more details.

**Comment by sharkdp** [comment]
I merged the documentation PR which seems uncontroversial. If someone disagrees with the decision here, let me know, and we can reopen this.
